subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/artifact/VariantResolvingArtifactSet.java

```diff
@@ -0,0 +1,186 @@
+import org.gradle.internal.component.model.ComponentArtifactResolveState;
+import org.gradle.internal.component.model.ComponentGraphResolveState;
+import org.gradle.internal.component.model.IvyArtifactName;
+import org.gradle.internal.component.model.VariantArtifactResolveState;
+import org.gradle.internal.component.model.VariantGraphResolveState;
+import org.gradle.internal.component.model.VariantResolveMetadata;
+import org.gradle.internal.lazy.Lazy;
+import org.gradle.internal.resolve.resolver.VariantArtifactResolver;
+
+import java.util.List;
+import java.util.function.Consumer;
+
+/**
+ * An {@link ArtifactSet} representing the artifacts contributed by a single variant in a dependency
+ * graph, in the context of the dependency referencing it.
+ */
+public class VariantResolvingArtifactSet implements ArtifactSet, VariantSelector.Factory {
+
+    private final VariantArtifactResolver variantResolver;
+    private final ComponentGraphResolveState component;
+    private final VariantGraphResolveState variant;
+    private final ComponentIdentifier componentId;
+    private final AttributesSchemaInternal schema;
+    private final ImmutableAttributes overriddenAttributes;
+    private final List<IvyArtifactName> artifacts;
+    private final ExcludeSpec exclusions;
+
+    private final Lazy<ImmutableSet<ResolvedVariant>> ownArtifacts = Lazy.locking().of(this::calculateOwnArtifacts);
+
+    public VariantResolvingArtifactSet(
+        VariantArtifactResolver variantResolver,
+        ComponentGraphResolveState component,
+        VariantGraphResolveState variant,
+        DependencyGraphEdge dependency
+    ) {
+        this.variantResolver = variantResolver;
+        this.component = component;
+        this.variant = variant;
+        this.componentId = component.getId();
+        this.schema = component.getMetadata().getAttributesSchema();
+        this.overriddenAttributes = dependency.getAttributes();
+        this.artifacts = dependency.getDependencyMetadata().getArtifacts();
+        this.exclusions = dependency.getExclusions();
+    }
+
+    @Override
+    public ResolvedArtifactSet select(Spec<? super ComponentIdentifier> componentFilter, VariantSelector selector, boolean selectFromAllVariants) {
+        if (!componentFilter.isSatisfiedBy(componentId)) {
+            return ResolvedArtifactSet.EMPTY;
+        } else {
+
+            if (selectFromAllVariants && !artifacts.isEmpty()) {
+                // Variants with overridden artifacts cannot be reselected since
+                // we do not know the "true" attributes of the requested artifact.
+                return new BrokenResolvedArtifactSet(new GradleException(
+                    String.format("Cannot reselect artifacts for variant %s of %s since it has an explicitly requested artifact. " +
+                        "Use a lenient ArtifactView to silence this error.", variant.getName(), componentId)
```

Comment:
💭 I understand what this is trying to say, since we've discussed it, but I am concerned that a build author seeing it for the first time might not.

"since it has an explicitly requested artifact" I think is misleading and this message should be talking the language of "declared dependency on an artifact" here, not "selected artifacts", since the dependency declarations in a build script is likely what will have to be investigated and changed to deal with this.  Also, why not include the artifact name, or the configuration creating the `ArtifactView`?  Probably the quickest way to help a build author jump to the source of problem.

This also runs into the issue with unnamed `ArtifactView`s, if there are multiple views for a configuration (javadoc and sources, perhaps), how do we know which one, or both, is broken?


I'd want to see a message like:

`Failed to select artifacts using view 'viewName' (if we had the name) of configuration 'myConf' because the dependency on 'compId' explicitly targets artifact 'artifactName'.  Views of configurations containing explicit artifact dependencies must use lenient=true.  LINK TO DOCS`.

Reply:
We worked together to revise this message.